### PR TITLE
Create a Key class to represent key data.

### PR DESCRIPTION
**Pull request by tedbow**

### CODE REVIEW

File: src/Key.php [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+     * @var string
+     */
+    private $scheme;
+
+    /**
+     * The public key value.
+     *
+     * @var string
+     */
+    private $public;
```
Comment (phenaproxima) [inline src/Key.php:33]: 😂😂

File: src/Metadata/RootMetadata.php [hunk 2]
```diff
@@ -70,19 +78,19 @@ public function getRoles(bool $allowUntrustedAccess = false): array
...
+     *   The keys for the metadata.
      */
-    public function getKeys(bool $allowUntrustedAccess = false): \ArrayObject
+    public function getKeys(bool $allowUntrustedAccess = false): array
     {
-        $this->ensureIsTrusted($allowUntrustedAccess);
-        return $this->getSigned()['keys'];
+        if (!isset($this->keys)) {
+            $this->ensureIsTrusted($allowUntrustedAccess);
+            $this->keys = [];
```
Comment (phenaproxima) [inline src/Metadata/RootMetadata.php:88]: Is it really necessary to store `$this->keys`? Feels like unnecessary state mutation in this class, which is probably something to be avoided if possible. That said, I don't feel strongly about it.

File: tests/KeyDBTest.php [hunk 3]
```diff
@@ -32,15 +31,12 @@ public function testCreateFromRootMetadata(): void
...
-        // Re-create the the key id which can be used by getKey().
-        $keyNormalized = JsonNormalizer::asNormalizedJson($key);
-        $keyId256 = hash('sha256', $keyNormalized);
-        self::assertSame($jsonEncodedKey, json_encode($keyDb->getKey($keyId256)));
-
-        // Ensure that changing a value in the key does not affect the internal state of the KeyDB object.
-        $key256['keyval']['new_key'] = 'new_value';
-        self::assertSame($jsonEncodedKey, json_encode($keyDb->getKey($keyId256)));
+        $keys = $rootMetadata->getKeys();
+        $key = array_shift($keys);
```
Comment (phenaproxima) [inline tests/KeyDBTest.php:35]:
```suggestion
        $key = reset($keys);
```

Nit: it might be wise to avoid state mutation where possible.

File: tests/Metadata/RootMetadataTest.php [hunk 4]
```diff
@@ -40,7 +40,6 @@ public function providerExpectedField(): array
 
         $data[] = ['signed:keys'];
         $firstKey = $this->getFixtureNestedArrayFirstKey('1.root.json', ['signed', 'keys']);
-        $data[] = ["signed:keys:$firstKey:keyid_hash_algorithms"];
```
Comment (phenaproxima) [inline tests/Metadata/RootMetadataTest.php:43]: Why was this removed?

File: tests/Metadata/TargetsMetadataTest.php [hunk 5]
```diff
@@ -88,7 +88,6 @@ public function providerExpectedField(): array
         $data[] = ['signed:delegations'];
         $data[] = ['signed:delegations:keys'];
         $firstKey = $this->getFixtureNestedArrayFirstKey($this->validJson, ['signed', 'delegations', 'keys']);
-        $data[] = ["signed:delegations:keys:$firstKey:keyid_hash_algorithms"];
```
Comment (phenaproxima) [inline tests/Metadata/TargetsMetadataTest.php:91]: Same question.

File: tests/Unit/KeyTest.php [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+use Tuf\Key;
+use PHPUnit\Framework\TestCase;
+
+/**
+ * @coversDefaultClass \Tuf\Key
+ */
+class KeyTest extends TestCase
+{
+
```
Comment (phenaproxima) [inline tests/Unit/KeyTest.php:13]:
```suggestion
```

File: tests/Unit/KeyTest.php [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+     * @covers ::createFromMetadata
+     * @covers ::getComputedKeyId
+     * @covers ::getPublic
+     * @covers ::getType
+     *
+     * @dataProvider providerCreateFromMetadata
+     */
+    public function testCreateFromMetadata(array $data): void
+    {
+        $defaultData = [
```
Comment (phenaproxima) [inline tests/Unit/KeyTest.php:24]:
```suggestion
        $data += [
```

File: tests/Unit/KeyTest.php [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+            'keytype' => 'ed11111',
+            'scheme' => 'scheme-ed11111',
+            'keyval' => new \ArrayObject(['public' => '12345']),
+        ];
+        $data += $defaultData;
```
Comment (phenaproxima) [inline tests/Unit/KeyTest.php:29]:
```suggestion
```

File: tests/Unit/KeyTest.php [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        self::assertSame('12345', $key->getPublic());
+        $keySortedCanonicalStruct = [
+            'keyid_hash_algorithms' => ['sha256', 'sha512'],
+            'keytype' => 'ed11111',
+            'keyval' => ['public' => '12345'],
+            'scheme' => 'scheme-ed11111',
+        ];
+        $keyCanonicalForm = json_encode($keySortedCanonicalStruct, JSON_UNESCAPED_SLASHES);
+
+        self::assertSame(hash('sha256', $keyCanonicalForm, false), $key->getComputedKeyId());
```
Comment (phenaproxima) [inline tests/Unit/KeyTest.php:42]:
```suggestion
        self::assertSame(hash('sha256', $keyCanonicalForm), $key->getComputedKeyId());
```

FALSE is the default third argument to `hash()`.

File: src/Metadata/RootMetadata.php (hunk 2 above)
Comment (tedbow) [reply src/Metadata/RootMetadata.php:88]: I added this because `RootMetadata` has no setters so you could never change the keys once the object was created was thinking of performance if `getKeys()` was called more than 1x. But right now it is not and this class is not final(maybe it should be) so we don't know a subclass won't add a setter for keys.

Removing

File: tests/Metadata/RootMetadataTest.php (hunk 4 above)
Comment (tedbow) [reply tests/Metadata/RootMetadataTest.php:43]: I realized the `\Tuf\Metadata\ConstraintsTrait::getKeyConstraints()` we had the validator for set up wrong. Check out the change there. Basically it should have been optional but it wasn't. It will be removed at some point in the Python library, see #174 and #179.

So this provider was checking that this field was *not* optional. It only passed because `getKeyConstraints()` was wrong. See `\Tuf\Tests\Unit\KeyTest::providerCreateFromMetadata()` that should we can validate now with or without this field.

File: tests/Metadata/TargetsMetadataTest.php (hunk 5 above)
Comment (tedbow) [reply tests/Metadata/TargetsMetadataTest.php:91]: same as above

File: tests/Unit/KeyTest.php (hunk 8 above)
Comment (tedbow) [reply tests/Unit/KeyTest.php:28]: fixed locally

File: tests/Unit/KeyTest.php (hunk 7 above)
Comment (tedbow) [reply tests/Unit/KeyTest.php]: fixed

**Review by phenaproxima — APPROVED** [review]
